src/main/scala/org/broadinstitute/dsde/firecloud/service/OAuthService.scala

```diff
@@ -26,189 +37,120 @@ object OAuthService {
-            case e: Exception => {
-              log.error("problem during OAuth code exchange", e)
-              requestContext.complete(StatusCodes.Unauthorized, e.getMessage)
-            }
-          }
-        } ~
-          parameter("error") { errorMsg =>
-            // echo the oauth error back to the user. Is that safe to do?
-            complete(StatusCodes.Unauthorized, errorMsg)
+    } catch {
+      case e: TokenResponseException =>
+        Future(RequestComplete(HttpResponse(StatusCodes.BadRequest, e.getContent)))
+    }
+  }
+
+  def getRefreshTokenStatus(userInfo: UserInfo): Future[PerRequestMessage] = {
+    val pipeline = addCredentials(userInfo.accessToken) ~> sendReceive
+    val tokenDateReq = Get(OAuthService.remoteTokenDateUrl)
+    val tokenDateFuture: Future[HttpResponse] = pipeline { tokenDateReq }
+    tokenDateFuture.map { response =>
+      response.status match {
+        case StatusCodes.OK =>
+          // rawls found a refresh token; check its date
+          val tokenDate = unmarshal[RawlsTokenDate].apply(response)
+          val howOld = Days.daysBetween(
+            DateTime.parse(tokenDate.refreshTokenUpdatedDate), DateTime.now)
+          howOld.getDays match {
+            case x if x < 90 =>
+              log.debug(s"User's refresh token is $x days old; all good!")
+              RequestComplete(StatusCodes.NoContent)
+            case x =>
+              log.info(s"User's refresh token is $x days old; requesting a new one.")
+              RequestComplete(StatusCodes.OK, Map("requiresRefresh" -> true))
           }
+        case StatusCodes.BadRequest =>
+          log.info(s"User has an illegal refresh token; requesting a new one.")
+          // rawls has a bad refresh token, restart auth.
+          RequestComplete(StatusCodes.OK, Map("requiresRefresh" -> true))
+        case StatusCodes.NotFound =>
+          log.info(s"User does not already have a refresh token; requesting a new one.")
+          // rawls does not have a refresh token for us. restart auth.
+          RequestComplete(StatusCodes.OK, Map("requiresRefresh" -> true))
+        case x =>
+          log.warn("Unexpected response code when querying rawls for existence of refresh token: "
+            + x.value + " " + x.reason)
+          RequestComplete(
+            StatusCodes.InternalServerError,
+            Map("error" -> Map("value" -> x.value, "reason" -> x.reason))
+          )
       }
     }
+  }
 
-
-  private def initiateAuth(state: String, approvalPrompt: String, requestContext: RequestContext) =  {
-    val gcsAuthUrl = HttpGoogleServicesDAO.getGoogleRedirectURI(state, approvalPrompt=approvalPrompt)
-    requestContext.redirect(gcsAuthUrl, StatusCodes.TemporaryRedirect)
+  private def updateNihStatus(subjectId: String, accessToken: String) = {
+    val userInfo = UserInfo("", OAuth2BearerToken(accessToken), -1, subjectId)
+    thurloeDao.getProfile(userInfo, updateExpiration = true)
```

Comment:
the fact that **get**Profile is the result of **update**NihStatus is concerning (see other comments on biz logic vs data access logic)